packages/analytics/shippers/elastic_v3/server/src/server_shipper.ts

```diff
@@ -0,0 +1,279 @@
+          });
+
+          if (!ok) {
+            throw new Error('Failed to connect to Elastic');
+          }
+
+          this.firstTimeOffline = null;
+          backoff = 1 * MINUTE;
+        }),
+        retryWhen((errors) =>
+          errors.pipe(
+            takeWhile(() => this.shutdown$.isStopped === false),
+            tap(() => {
+              if (!this.firstTimeOffline) {
+                this.firstTimeOffline = Date.now();
+              } else if (Date.now() - this.firstTimeOffline > 24 * HOUR) {
+                this.internalQueue.length = 0;
+              }
+              backoff = backoff * 2;
+              if (backoff > 1 * HOUR) {
+                backoff = 1 * HOUR;
+              }
+            }),
+            delayWhen(() => timer(backoff))
+          )
+        )
+      )
+      .subscribe();
+  }
+
+  private setInternalSubscriber() {
+    // Check the status of the queues every 1 second.
+    const subscription = merge(
+      interval(1000),
+      // Using a promise because complete does not emit through the pipe.
+      from(firstValueFrom(this.shutdown$, { defaultValue: true }))
+    )
+      .pipe(
+        // Only move ahead if it's opted-in and online.
+        filter(() => this.isOptedIn === true && this.firstTimeOffline === null),
+
+        // Send the events now if (validations sorted from cheapest to most CPU expensive):
+        // - We are shutting down.
+        // - There are some events in the queue, and we didn't send anything in the last 10 minutes.
+        // - The last time we sent was more than 10 seconds ago and:
+        //   - We reached the minimum batch size of 10kB per request in our leaky bucket.
+        //   - The queue is full (meaning we'll never get to 10kB because the events are very small).
+        filter(
+          () =>
+            (this.internalQueue.length > 0 &&
+              (this.shutdown$.isStopped ||
+                Date.now() - this.lastBatchSent$.value >= 10 * MINUTE)) ||
+            (Date.now() - this.lastBatchSent$.value >= 10 * SECOND &&
+              (this.internalQueue.length === 1000 ||
+                this.getQueueByteSize(this.internalQueue) >= 10 * KIB))
+        ),
+
+        // Send the events
+        exhaustMap(async () => {
+          this.lastBatchSent$.next(Date.now());
+          const eventsToSend = this.getEventsToSend();
+          await this.sendEvents(eventsToSend);
+        })
+      )
+      .subscribe(() => {
+        if (this.shutdown$.isStopped) {
+          subscription.unsubscribe();
+        }
+      });
+  }
+
+  /**
+   * Calculates the size of the queue in bytes.
+   * @returns The number of bytes held in the queue.
+   */
+  private getQueueByteSize(queue: Event[]) {
+    return queue.reduce((acc, event) => {
+      return acc + Buffer.from(JSON.stringify(event)).length;
+    }, 0);
+  }
+
+  /**
+   * Returns a queue of events of up-to 10kB.
+   * @private
+   */
+  private getEventsToSend(): Event[] {
```

Comment:
NIT: the function has a side effect of mutating the `internalQueue` value. I would make that explicit, probably by naming it something like `popEventsToSend`, or at least documenting it in the jsdoc.